controllers/raycluster_controller.go

```diff
@@ -0,0 +1,424 @@
+func (r *RayClusterReconciler) updateStatus(ctx context.Context, rc *dcv1alpha1.RayCluster) error {
+	mNodes, err := r.modifyStatusNodes(ctx, rc)
+	if err != nil {
+		return fmt.Errorf("cannot modify cluster status nodes: %w", err)
+	}
+
+	mWorkedFields, err := r.modifyStatusWorkerFields(ctx, rc)
+	if err != nil {
+		return fmt.Errorf("cannot modify cluster status worker fields: %w", err)
+	}
+
+	if mNodes || mWorkedFields {
+		if err = r.Status().Update(ctx, rc); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// modifyStatusNodes will ensure the status contains an accurate list of all the pods in the cluster.
+func (r *RayClusterReconciler) modifyStatusNodes(ctx context.Context, rc *dcv1alpha1.RayCluster) (bool, error) {
+	podList := &corev1.PodList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(rc.Namespace),
+		client.MatchingLabels(ray.MetadataLabels(rc)),
+	}
+	if err := r.List(ctx, podList, listOpts...); err != nil {
+		return false, fmt.Errorf("cannot list ray pods: %w", err)
+	}
+
+	var podNames []string
+	for _, pod := range podList.Items {
+		podNames = append(podNames, pod.Name)
+	}
+	sort.Strings(podNames)
+
+	if reflect.DeepEqual(podNames, rc.Status.Nodes) {
+		return false, nil
+	}
+
+	log := r.getLogger(ctx)
+
+	log.V(1).Info("modifying status", "path", ".status.nodes", "value", podNames)
+	rc.Status.Nodes = podNames
+
+	return true, nil
+}
+
+// modifyStatusWorkerFields syncs certain worker deployment fields into the status.
+func (r *RayClusterReconciler) modifyStatusWorkerFields(ctx context.Context, rc *dcv1alpha1.RayCluster) (bool, error) {
+	worker, err := ray.NewDeployment(rc, ray.ComponentWorker)
+	if err != nil {
+		return false, err
+	}
+
+	err = r.Get(ctx, client.ObjectKeyFromObject(worker), worker)
+	if client.IgnoreNotFound(err) != nil {
+		return false, err
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(worker.Spec.Selector)
+	if err != nil {
+		return false, err
+	}
+
+	log := r.getLogger(ctx)
+
+	var modified bool
+	if rc.Status.WorkerSelector != selector.String() {
+		rc.Status.WorkerSelector = selector.String()
+		modified = true
+
+		log.V(1).Info("modifying status", "path", ".status.workerSelector", "value", rc.Status.WorkerSelector)
+	}
+	if rc.Status.WorkerReplicas != *worker.Spec.Replicas {
+		rc.Status.WorkerReplicas = *worker.Spec.Replicas
+		modified = true
+
+		log.V(1).Info("modifying status", "path", ".status.workerReplicas", "value", rc.Status.WorkerReplicas)
+	}
+
+	return modified, nil
+}
+
+type loggerKeyType int
+
+const loggerKey loggerKeyType = iota
+
+func (r *RayClusterReconciler) setLogger(ctx context.Context, logger logr.Logger) (context.Context, logr.Logger) {
+	return context.WithValue(ctx, loggerKey, logger), logger
+}
+
+func (r *RayClusterReconciler) getLogger(ctx context.Context) logr.Logger {
+	if ctx == nil {
+		return r.Log
+	}
+	if ctxLogger, ok := ctx.Value(loggerKey).(logr.Logger); ok {
+		return ctxLogger
+	}
+
+	return r.Log
+}
```

Comment:
Is there no `ContextWithLogger` type we can leverage?

Reply:
Ha! I was working on adding that this morning. Check out https://github.com/dominodatalab/distributed-compute-operator/pull/2/commits/1efdd6b56eb95f35355d1c7bf7eaf205ca6a4825.